### ISSUE TITLE
Using string enums with same casing of keys and values

### ISSUE THREAD

**Issue by stekycz** [mikro-orm/mikro-orm#359]
I want to use a custom enum. However, I do not want to use numeric values in data because:

1. they might differ when changing the order of values in TS enum
2. they are not readable in data itself because I need to know the meaning and number mapping

I also want to use enum keys to be always upper case like
```ts
enum PublisherType {
    GLOBAL = 'GLOBAL',
    LOCAL = 'LOCAL'
}
```
I have used inspiration from [your own test](https://github.com/mikro-orm/mikro-orm/blob/master/tests/entities-sql/Publisher2.ts#L6-L9).

However, I am not able to do that because [the `@Enum` decorator uses values that are different from its value](https://github.com/mikro-orm/mikro-orm/blob/master/lib/metadata/MetadataDiscovery.ts#L481-L483). I do not know the reason why it is that way, however, it seems as not needed constraint.

I have seen that extra parameter `() => PublisherType` is needed after the latest changes which duplicate the type definition that is defined in TS. However, that is something I can live with.

I would expect the enum uses all values defined in the `Enum`.

**Comment by B4nan** [comment]
> However, I am not able to do that because the @Enum decorator uses values that are different from its value. I do not know the reason why it is that way, however, it seems as not needed constraint.

Hmm I think I understand, this use case is not covered. The reason for the highlighted code is that numeric enums will also have reverse mapping, which this particular code tries to remove. 

```typescript
enum PublisherType {
  LOCAL = 1,
  GLOBAL = 2,
}

const keys = Object.keys(PublisherType);
console.log(keys); // ["1", "2", "LOCAL", "GLOBAL"]
const items = Object.values(PublisherType).filter(val => !keys.includes(val));
console.log(items); // [1, 2]
```

Should be easy to get around this. In the meantime, you could probably get around this by using custom type instead. Not sure if providing the items manually will help (`@Enum({ items: [...] })`), as that might be overridden by the method you linked (which is another bug if so, will need to try).

> I have seen that extra parameter () => PublisherType is needed after the latest changes which duplicate the type definition that is defined in TS. However, that is something I can live with.

This is needed only if you do not use the default ts-morph metadata provider. All SQL tests are using reflect metadata provider where this is needed.

**Comment by B4nan** [comment]
Or you could obviously just rename your enum keys (add some suffix, etc) for now and once fixed, rename them back once the fix is merged.

**Comment by B4nan** [comment]
Fixed in `3.3.1`.
